engine/runtime/src/main/java/org/enso/interpreter/runtime/TruffleCompilerContext.java

```diff
@@ -278,29 +284,363 @@ public String formatDiagnostic(
+                    SuggestionBuilder.apply(module, compiler)
+                        .build(module.getName(), module.getIr())
+                        .toVector()
+                        .filter(Suggestion::isGlobal);
+                var exports = exportsBuilder.build(module.getName(), module.getIr());
+                exportsMap.addAll(module.getName(), exports);
+                return sug;
+              })
+          .map(
+              suggestion -> {
+                var reexport = exportsMap.get(suggestion).map(s -> s.toString());
+                return suggestion.withReexport(reexport);
+              })
+          .foreach(suggestions::add);
+
+      var cachedSuggestions =
+          new SuggestionsCache.CachedSuggestions(
+              libraryName,
+              new SuggestionsCache.Suggestions(suggestions),
+              context
+                  .getPackageRepository()
+                  .getPackageForLibraryJava(libraryName)
+                  .map(p -> p.listSourcesJava()));
+      var cache = SuggestionsCache.create(libraryName);
+      var file = saveCache(cache, cachedSuggestions, useGlobalCacheLocations);
+      return file.isPresent();
+    } catch (Throwable e) {
+      logSerializationManager(
+          Level.SEVERE,
+          "Serialization of suggestions `" + libraryName + "` failed: " + e.getMessage() + "`",
+          e);
+      throw e;
+    }
+  }
+
+  public scala.Option<List<org.enso.polyglot.Suggestion>> deserializeSuggestions(
+      LibraryName libraryName) throws InterruptedException {
+    var option = deserializeSuggestionsImpl(libraryName);
+    return option.map(s -> s.getSuggestions());
+  }
+
+  private scala.Option<SuggestionsCache.CachedSuggestions> deserializeSuggestionsImpl(
+      LibraryName libraryName) throws InterruptedException {
+    var pool = serializationPool;
+    if (pool.isWaitingForSerialization(toQualifiedName(libraryName))) {
+      pool.abort(toQualifiedName(libraryName));
+      return scala.Option.empty();
+    } else {
+      pool.waitWhileSerializing(toQualifiedName(libraryName));
+      var cache = SuggestionsCache.create(libraryName);
+      var loaded = loadCache(cache);
+      if (loaded.isPresent()) {
+        logSerializationManager(Level.FINE, "Restored suggestions for library [{0}].", libraryName);
+        return scala.Option.apply(loaded.get());
+      } else {
+        logSerializationManager(
+            Level.FINE, "Unable to load suggestions for library [{0}].", libraryName);
+        return scala.Option.empty();
+      }
+    }
+  }
+
+  scala.Option<ImportExportCache.CachedBindings> deserializeLibraryBindings(LibraryName libraryName)
+      throws InterruptedException {
+    var pool = serializationPool;
+    if (pool.isWaitingForSerialization(toQualifiedName(libraryName))) {
+      pool.abort(toQualifiedName(libraryName));
+      return scala.Option.empty();
+    } else {
+      pool.waitWhileSerializing(toQualifiedName(libraryName));
```

Comment:
is this safe? nothing stops context from serializing the library in between waiting and loading the cache? Not that it is likely to happen, but stil.

Reply:
It is as _"safe"_ as the original implementation - I haven't changed the logic. True, the original code was probably not really multithreading ready. The fact that it was using `Thread.sleep` for coordination of threads speaks for itself.

I haven't attempted to make the code bulletproof. That would require bigger rewrite. Btw. can the code even be really safe when multiple processes access the caches at the same time? That might require some co-ordination with filesystem. What goals do we have for the caching infrastructure? 

My current implementation implies:
- caches are written down in a background executor thread
- reading the caches happens in a caller thread and waits until the background thread finishes

If we want more, we should probably sit down and collect other requirements/constraints. Then rewrite the implementation.

Reply:
It would be great if the coordination of _filesystem access_ to caches was encapsulated in a single class like  `SerializationPool`. To replace this `waitWhileSerializing` method we would need something like:
```java
public <T> T readCaches(QualifiedName name, Callbable<T> action) {...}
```
that would prevent serialization of `name` key while executing the action. I can try to redesign that.